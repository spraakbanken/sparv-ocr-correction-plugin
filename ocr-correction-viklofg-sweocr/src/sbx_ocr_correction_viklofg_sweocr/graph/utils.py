import re
from typing import List, Tuple, TypedDict

import diff_match_patch as dmp_module

dmp = dmp_module.diff_match_patch()

ENDING_WHITESPACE = re.compile(r"\s$")


def end_with_space(s: str) -> str:
    if not s:
        return s
    # print(f"{s[-1]=}")
    # print(f"{ENDING_WHITESPACE.fullmatch(s[-1])=}")
    return f"{s} " if (ENDING_WHITESPACE.fullmatch(s[-1]) is None) else s


def token_diff(s1: str, s2: str) -> List[Tuple[int, str]]:
    d = dmp.diff_main(s1, s2)
    dmp.diff_cleanupSemantic(d)
    return d


EditRange = TypedDict("EditRange", {"from": int, "to": int, "insert": str})


def edit_range(s0: str, s: str) -> EditRange:
    """
    >>> edit_range('0123456789', '0189')
    {from: 2, to: 8, insert: ''}

    edit_range('0123456789', '01') // => {from: 2, to: 10, insert: ''}
    edit_range('0123456789', '89') // => {from: 0, to: 8, insert: ''}
    edit_range('0123456789', '') // => {from: 0, to: 10, insert: ''}

    edit_range('0123456789', '01xyz89') // => {from: 2, to: 8, insert: 'xyz'}
    edit_range('0123456789', '01xyz') // => {from: 2, to: 10, insert: 'xyz'}
    edit_range('0123456789', 'xyz89') // => {from: 0, to: 8, insert: 'xyz'}
    edit_range('0123456789', 'xyz') // => {from: 0, to: 10, insert: 'xyz'}

    edit_range('', '01') // => {from: 0, to: 0, insert: '01'}
    """
    patches = token_diff(s0, s)  # noqa: F841
    # const patches = token_diff(s0, s)
    # const pre = R.takeWhile<[number, string]>(i => i[0] == 0, patches)
    # const post = R.takeLastWhile<[number, string]>(i => i[0] == 0, R.drop(pre.length, patches))  # noqa: E501
    # const from = pre.map(i => i[1]).join('').length
    # const postlen = post.map(i => i[1]).join('').length
    # const to = s0.length - postlen
    # const insert = s.slice(from, s.length - (s0.length - to))
    # return {from, to, insert}


def uniq(xs: List[str]) -> List[str]:
    used = set()
    return [x for x in xs if x not in used and (used.add(x) or True)]
