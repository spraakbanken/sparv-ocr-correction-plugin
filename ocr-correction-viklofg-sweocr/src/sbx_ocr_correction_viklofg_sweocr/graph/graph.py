import enum
from typing import Dict, List, Optional, TypedDict, TypeVar

from sbx_ocr_correction_viklofg_sweocr.graph import token, utils
from sbx_ocr_correction_viklofg_sweocr.graph.token import Token

A = TypeVar("A")


class Side(enum.StrEnum):
    source = "source"
    target = "target"


class SourceTarget(TypedDict):
    source: List[Token]
    target: List[Token]


class Edge(TypedDict):
    # a copy of the identifier used in the edges object of the graph
    id: str
    # these are ids to source and target tokens
    ids: List[str]
    # labels on this edge
    labels: List[str]
    # is this manually or automatically aligned
    manual: bool
    comment: Optional[str] = None


Edges = dict[str, Edge]


class Graph(SourceTarget):
    edges: Edges
    comment: Optional[str] = None


def edge(
    ids: List[str],
    labels: List[str],
    *,
    comment: Optional[str] = None,
    manual: bool = False,
) -> Edge:
    ids_sorted = sorted(ids)
    labels_nub = utils.uniq(labels)
    return Edge(
        id=f"e-{'-'.join(ids_sorted)}",
        ids=ids_sorted,
        labels=labels_nub,
        manual=manual,
        comment=comment,
    )


def edge_record(es: List[Edge]) -> Dict[str, Edge]:
    return {e["id"]: e for e in es}


def init(s: str, *, manual: bool = False) -> Graph:
    return init_from(token.tokenize(s), manual=manual)


def init_from(tokens: List[str], *, manual: bool = False) -> Graph:
    return {
        "source": token.identify(tokens, "s"),
        "target": token.identify(tokens, "t"),
        "edges": edge_record(
            (edge([f"s{i}", f"t{i}"], [], manual=manual) for i, _ in enumerate(tokens))
        ),
    }


def unaligned_set_side(g: Graph, side: Side, text: str) -> Graph:
    text0 = get_side_text(g, side)
    from_, to = utils.edit_range(text0, text)
    new_text = text[from_ : (len(text) - (len(text0) - to))]  # noqa: F841
    # return unaligned_modify(g, from_, to, new_text, side)


def get_side_text(g: Graph, side: Side) -> str:
    return token.text(g[side])
