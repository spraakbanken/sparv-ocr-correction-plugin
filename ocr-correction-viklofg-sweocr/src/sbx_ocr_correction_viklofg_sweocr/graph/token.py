import re
from typing import List, TypedDict

from sbx_ocr_correction_viklofg_sweocr.graph import utils


class Text(TypedDict):
    text: str


class Token(Text, TypedDict):
    id: str


class Span(TypedDict):
    begin: int
    end: int


def text(ts: List[Text]) -> str:
    """The text in some tokens

    >>> text(identify(tokenize('apa bepa cepa '), '#'))
    'apa bepa cepa '

    """
    return "".join(texts(ts))


def texts(ts: List[Text]) -> List[str]:
    """The texts in some tokens

    >>> texts(identify(tokenize('apa bepa cepa '), '#'))
    ['apa ', 'bepa ', 'cepa ']
    """
    return [t["text"] for t in ts]


def tokenize(s: str) -> List[str]:
    """Tokenizes text on whitespace, prefers to have trailing whitespace."""
    return list(
        map(
            utils.end_with_space,
            re.findall(r"\s*\S+\s*", s) or re.findall(r"^\s+$", s) or [],
        )
    )


def identify(toks: List[str], prefix: str) -> List[Token]:
    return [Token(text=text, id=f"{prefix}{i}") for i, text in enumerate(toks)]
