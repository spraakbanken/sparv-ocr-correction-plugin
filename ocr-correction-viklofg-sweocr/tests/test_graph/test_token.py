from typing import List

import pytest
from sbx_ocr_correction_viklofg_sweocr.graph.token import Token, identify, tokenize


def test_can_create_token() -> None:
    token = Token(text="a text", id="s0")

    assert token["id"] == "s0"
    assert token["text"] == "a text"


@pytest.mark.parametrize(
    "text, expected",
    [
        ("", []),
        (" ", [" "]),
        ("    ", ["    "]),
        ("apa bepa cepa", ["apa ", "bepa ", "cepa "]),
        ("  apa bepa cepa", ["  apa ", "bepa ", "cepa "]),
        ("  apa bepa cepa  ", ["  apa ", "bepa ", "cepa  "]),
    ],
)
def test_tokenize(text: str, expected: List[str], snapshot) -> None:
    actual = tokenize(text)

    assert actual == expected
    assert actual == snapshot


def test_identify() -> None:
    assert identify(["apa", "bepa"], "#") == [
        {"text": "apa", "id": "#0"},
        {"text": "bepa", "id": "#1"},
    ]
