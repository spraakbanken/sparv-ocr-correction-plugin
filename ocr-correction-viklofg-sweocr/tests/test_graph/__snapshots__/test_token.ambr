# name: test_tokenize[    -expected2]
  <class 'list'> [
    '    ',
  ]
---
# name: test_tokenize[  apa bepa cepa  -expected5]
  <class 'list'> [
    '  apa ',
    'bepa ',
    'cepa  ',
  ]
---
# name: test_tokenize[  apa bepa cepa-expected4]
  <class 'list'> [
    '  apa ',
    'bepa ',
    'cepa ',
  ]
---
# name: test_tokenize[ -expected1]
  <class 'list'> [
    ' ',
  ]
---
# name: test_tokenize[-expected0]
  <class 'list'> [
  ]
---
# name: test_tokenize[apa bepa cepa-expected3]
  <class 'list'> [
    'apa ',
    'bepa ',
    'cepa ',
  ]
---
